Utility classes to make it easier to use sandbox facet API for most common cases

epotyom opened the pull request:
In the initial sandbox facet module PR @gsmiller [suggested](https://github.com/apache/lucene/pull/13568#issuecomment-2249005915) adding helpers to make common tasks easier.

This implementation uses existing type `FacetResult` to return results. The type has limitations, e.g. it supports only one value per label. At the same time, it is simple, users are already familiar with it, and it is probably good enough for the most common tasks. Users who need more complex results can use the lower level API FacetCutter, FacetRecorder, etc to get multiple aggregations per label, to implement custom sort order or any other non-trivial faceting task.

It's a draft, if people are ok with the approach overall I will add more unit tests and javadocs.

epotyom committed:
Initial commit for utility classes for the sandbox facets module

Simplifies requesting taxonomy and long range facets
Makes sure we don't count docs two times for the same index field when there are multiple requests
Returns FacetResult - to make it easier to migrate for classic facets module users

TODOs:
- better names - maybe use names suggested in TK
- implement for all existing facet cutters
- add comprehensive tests
- fix total value where possible in tests

epotyom committed:
TMP: DrillSidewaysFacetOrchestrator.

epotyom committed:
TMP: DrillSidewaysFacetOrchestrator part 2

epotyom committed:
Make FacetBuilder abstract class

epotyom committed:
tidy, licence and other small changes to pass gradlew check

epotyom committed:
Some FacetOrchestrator and DrillSidewaysFacetOrchestrator improvements: reuse some code, make more abstract

epotyom committed:
add CommonFacetBuilder, some unit tests and small canges

epotyom committed:
Minor changes

stefanvodita commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/facet/utils/CommonFacetBuilder.java:
```diff
@@ -0,0 +1,58 @@
+  }
+
+  @Override
+  FacetCutter createFacetCutter() {
```
Would it be a problem that this is called `create`, giving the impression that we get a new cutter? Reusing the old one could be problematic, couldn't it, if the user expects a new cutter?

epotyom replied:
Ah you are right, this method doesn't have to create FacetCutter, e.g. FacetCutter might be created in advance and returned by the method. Renamed to `getFacetCutter`

stefanvodita commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/facet/utils/BaseFacetBuilder.java:
```diff
@@ -0,0 +1,155 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.lucene.sandbox.facet.utils;
```
Should we have a package for the facet builders?

epotyom replied:
I want to keep `*FacetOrchstrator`s in the same package as `FacetBuilder`s, so that we can keep some `FacetBuilder` method package-private. So the only file that maybe doesn't belong in the same package is `ComparableUtils.java`. At the same time the idea of ComparableUtils is similar - make it easier for users to use the new module, so I thought we could reduce number of packages and keep meaningful package structure at the same time... But I don't feel strongly about it. WDYT?

stefanvodita replied:
I get your argument. Let's keep it as is then.

stefanvodita reviewed (commented):
I had a glance at everything and didn't see anything objectionable. It's great if we're making this module easier to use. I have some non-blocking concerns that this is a lot of new code, with more abstractions, and it doesn't have a user waiting to use it. There are arguments against those concerns as well and I personally like the new aggregations, but I thought it was worth raising the point.

epotyom committed:
private LongRangeFacetBuilder ctor

epotyom commented:
Thanks for reviewing @stefanvodita !

> I have some non-blocking concerns that this is a lot of new code, with more abstractions, and it doesn't have a user waiting to use it. There are arguments against those concerns as well and I personally like the new aggregations, but I thought it was worth raising the point.

I agree with the concern, at the same time it might be one of these chicken and egg cases when users avoid using the feature just because it doesn't have simple enough API for simple use cases?

One use case for the new utils I have in mind is luceneutil perf tests - it puts me off that [we need 90 lines](https://github.com/mikemccand/luceneutil/blob/dac76d58c579bf440ff4c339622c236202ec65d0/src/main/perf/SearchTask.java#L261-L352) to test performance of the new API, while for existing facets module it is [only two lines](https://github.com/mikemccand/luceneutil/blob/dac76d58c579bf440ff4c339622c236202ec65d0/src/main/perf/SearchTask.java#L403-L404).

epotyom commented on lucene/demo/src/java/org/apache/lucene/demo/facet/SandboxFacetsExample.java:
```diff
@@ -130,6 +135,88 @@ void index() throws IOException {
 
+  /**
+   * Example for {@link FacetBuilder} usage - simple API that provides results in a format very
+   * similar to classic facets module. It doesn't give all flexibility available with {@link
```
Good idea, I'll do that.

Shradha26 replied:
Thanks for creating the `FacetBuilder` and `FacetOrchestrator` abstractions - they do really simplify the usage of the new API.

epotyom committed:
Mention sandbox facet module in facets userguide

epotyom committed:
Rename BaseFacetBuilder#createFacetCutter to #getFacetCutter

stefanvodita reviewed (approved):
Thank you for addressing my comments! I'm up for merging this, while keeping in mind the concerns we discussed.

Shradha26 commented on lucene/demo/src/java/org/apache/lucene/demo/facet/SandboxFacetsExample.java:
```diff
@@ -130,6 +135,88 @@ void index() throws IOException {
+        .collect(query, ds);
+
+    //// Results
+    FacetResult autorResults = authorFacetBuilder.getResult();
```
[nit] authorResults instead of autorResults?

Shradha26 commented on lucene/demo/src/java/org/apache/lucene/demo/facet/SandboxFacetsExample.java:
```diff
@@ -130,6 +135,88 @@ void index() throws IOException {
+            "Price",
+            new LongRange("0-10", 0, true, 10, true),
+            new LongRange("10-20", 10, true, 20, true));
+
```
To maintain output consistency between different examples, we could add something equivalent to what we have in `simpleFacetsWithSearch `? Or, we could remove it from there. 
```
System.out.println(
        "Search results: totalHits: "
            + topDocs.totalHits
            + ", collected hits: "
            + topDocs.scoreDocs.length);
```

epotyom replied:
I did this intentionally, to demonstrate that there are methods to do facet only and facets with search (examples that have`WithSearch` suffix also compute/print TopDocs). We can rename this example to `simpleFacetsOnlyWithDrillSideways` maybe?

Shradha26 commented on lucene/demo/src/java/org/apache/lucene/demo/facet/SandboxFacetsExample.java:
```diff
@@ -130,6 +135,88 @@ void index() throws IOException {
+
+    //// Search and collect
+    TopDocs topDocs =
+        new FacetOrchestrator()
```
How do we handle cases where more than one `FacetBuilder` instances specify topN?

epotyom replied:
Each `FacetBuilder` can have its own topN value set by `withTopN()`, which is used to compute `FacetBuilder#getResult`

Shradha26 commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/facet/utils/LongRangeFacetBuilder.java:
```diff
@@ -0,0 +1,51 @@
+import org.apache.lucene.sandbox.facet.labels.RangeOrdToLabel;
+
+/** {@link FacetBuilder} factory for long range facets. */
+public final class LongRangeFacetBuilder {
```
Why don't we make `LongRangeFacetBuilder` extend `CommonFacetBuilder`? The `create` method returning something of a different type is a bit confusing to me.

epotyom replied:
The idea is that CommonFacetBuilder can be used to compute results for long range facets, so there is no need to create a new FacetBuilder subclass. But I thought it would be nice to have a utility class (I should have added a private constructor to make it static-only, will do) that helps to get CommonFacetBuilder instance.

I can remove the class and let users rely on examples, or rename it to `RangeFacetBuilder` and add similar methods for `double` ranges. WDYT?

Shradha26 commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/facet/utils/CommonFacetBuilder.java:
```diff
@@ -0,0 +1,58 @@
+ * Common {@link FacetBuilder} that works with any {@link FacetCutter} and {@link OrdToLabel}
+ * provided by user.
+ */
+public final class CommonFacetBuilder extends BaseFacetBuilder<CommonFacetBuilder> {
```
I'm not sure when I should be using `CommonFacetBuilder ` vs `BaseFacetBuilder`? Some examples, even in the documentation would help? For e.g, why can't `TaxonomyFacetBuilder` extend `CommonFacetBuilder`?

epotyom replied:
BaseFacetBuilder is abstract package-private class that implements common functionality that different FacetBuilder implementations can reuse.

CommonFacetBuilder is a final class, kind of generic facet builder that can be used with any FacetCutter, but as a consequence it lacks some facet type specific functionality. E.g. using it for taxonomy facets makes sense only if we don't need to specify parent path and don't need to compute overall value. The idea is that creating FacetBuilder should not be required for every facet type, just for the types that require some custom logic, such as taxonomy facets.

WDYT?

epotyom committed:
Improve javadocs and fix some typos

stefanvodita commented:
@Shradha26, if you don't have more feedback, I'll merge this soon.

epotyom committed:
Changes from offline discussion with Shradha:
- Improved package documentation for both sandbox/facets and sandbox/facets/utils
- Renamed LongRangeFacetBuilder to RangeFacetBuilderFactory and added factory methods for double range facets
- Implemented LongValueFacetBuilder
- Fixed small bugs in TaxonomyFacetBuilder and SandboxFacetsExample
- Improved unit tests

epotyom committed:
CHANGES.txt record

stefanvodita reviewed (commented):
I'm happy with it, I'm just wondering if you want to put it in 10.2.

epotyom committed:
Changes to pass lint check for JDK 23; will have to be rolled back in 10.2 backport

epotyom commented:
> I'm happy with it, I'm just wondering if you want to put it in 10.2.

Makes sense, I updated the branch.

Also rebased once again, and added a commit https://github.com/apache/lucene/pull/14237/commits/b5c74c633880f290348ee72f4351d869b827c013 to pass lint test for the new min required JDK version 23. I think it has to be rolled back for 10.2 backport?
